### PR TITLE
Add source of truth prop

### LINKED ISSUE #22: Add ability to control selected items through selected{Labels,Values}
#13 resulted in the suboptimal situation that if you're only interested in primitive labels/values for the selected items, you can get those through

```svelte
<MultiSelect bind:selectedLabels bind:selectedValues />
```

but it's no longer a two-way binding. You cannot control the component's internal `selected` array through those props since they are reactive to `selected` internally:

```svelte
$: selectedLabels = selected.map((op) => op.label)
$: selectedValues = selected.map((op) => op.value)
```

Need to figure out how to allow users to reverse this dependence and set `bind:selectedLabels/Values` as the source of truth.

### PR DESCRIPTION
**Pull request by GodPingu**
This pull resolves #22 by adding a sourceOfTruth prop which is used by the multiselect component to determine which selected prop to respond to changes on. I also renamed selected to selectedOptions for clarity in relation to the SourceOfTruth enum.

This pull request includes a new prop, a new enum, a new example, as well as changing the name of "selected" to "selectedOptions"

### CODE REVIEW

File: src/lib/MultiSelect.svelte [hunk 1]
```diff
@@ -1,21 +1,34 @@
...
   export let options: ProtoOption[]
-  export let selected: Option[] =
-    (options as Option[]).filter((op) => op?.preselected) ?? []
+
+  export let selectedLabels: Primitive[] = []
+  export let selectedValues: Primitive[] = []
+  export let selectedOptions: Option[] = []
+
+  export let sourceOfTruth: SourceOfTruth = SourceOfTruth.options
+  switch (sourceOfTruth) {
```
Comment (janosh) [inline src/lib/MultiSelect.svelte:21]: Not a huge fan of switch cases tbh as I find them less readable and more verbose. Could we swap this for `if`/`else if`/...?

File: src/lib/index.ts [hunk 2]
```diff
@@ -13,6 +13,12 @@ export type Option = {
   [key: string]: unknown // allow any other keys users might want
 }
 
+export enum SourceOfTruth {
```
Comment (janosh) [inline src/lib/index.ts:16]: I think it's safe enough to type-hint the `sourceOfTruth` prop as `'options' | 'labels' | 'values'`. No need to use a enum here I think. TS will complain any way if we make a non-sensical comparison like ``case `foobar`:``.

**Review by janosh — CHANGES_REQUESTED** [review]

**Comment by janosh** [comment]
@GodPingu Thanks for the contribution! Could you document the new `sourceOfTruth` prop in the readme and fix the other failing test? I also left some comments.

File: src/lib/index.ts (hunk 2 above)
Comment (GodPingu) [reply src/lib/index.ts]: Agree about this I've just switched it to a type over an enum

File: src/lib/MultiSelect.svelte (hunk 1 above)
Comment (GodPingu) [reply src/lib/MultiSelect.svelte]: Swapped these over

**Comment by GodPingu** [comment]
Believe I've fixed the test and completed any changes you wanted

**Comment by janosh** [comment]
Thanks for the getting the ball rolling. After giving it some more thought, the easiest solution actually seems to not force options to be objects internally but rather to work with whatever the user passes in. If we get strings or numbers, just leave them as strings/numbers. I wasn't sure how well that would work but after giving it a try locally, seems like it requires much less code and is more readable. It still doesn't allow controlling the selected options with two-way binding via `selectedLabels` and `selectedValues` but those were mostly added for people who didn't want to deal with options objects if they supplied simple strings. Now that this restriction is lifted, I think it is fine to mark `selectedLabels` and `selectedValues` as read-only since bind:selected can now be used to control options externally in the same shape as they were provided.

I'll push my changes here and then merge your PR so your work is credited.

**Comment by janosh** [comment]
Oops, I messed this up. Accidentally force pushed to your fork. Very sorry! Could you open another PR?

**Comment by GodPingu** [comment]
Just created a new pull request thanks

**Comment by GodPingu** [comment]
Having given thought to it I think this makes sense it was getting quite verbose trying to deal with values/labels etc. I think the use of slot components where users need the difference between values and labels/display info, as already implemented, is the best solution

**Pull request closed without merging**